### ISSUE TITLE
Unignoring directories and their inner files in .eslintignore does not work

### ISSUE THREAD

**Issue by belochub** [eslint/eslint#11684]
**Tell us about your environment**

* **ESLint Version:** 5.16.0
* **Node Version:** 10.14.2
* **npm Version:** 6.4.1

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
  "rules": { "no-octal": "error" }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

A minimal example that can be used to reproduce the issue:
In the project with the following directory structure:
```
.
├── dir
│   └── file.js
├── .eslintignore
├── .eslintrc
├── file.js
├── node_modules
│   └── ...
├── package.json
└── package-lock.json
```
This is the `.eslintignore` file:
```
*
!/dir/
!/dir/file.js
```

Both files `./file.js` and `./dir/file.js` contain the following code:

```js
01
```

```bash
npx eslint .
```

**What did you expect to happen?**

File `./dir/file.js` is linted and an error is reported:
```
/path/to/project/dir/file.js
  1:1  error  Octal literals should not be used  no-octal

✖ 1 problem (1 error, 0 warnings)
```

**What actually happened? Please include the actual, raw output from ESLint.**

```
All files matched by '.' are ignored.
```

**Are you willing to submit a pull request to fix this bug?**
Yes, I have a possible fix for this issue that I would like to submit.

AFAIU, this was broken in https://github.com/eslint/eslint/pull/6783 since I've tried this on the ESLint versions that do not include that change (`<3.4.0`), and they seem to be working correctly.

**Comment by platinumazure** [comment]
Hi @belochub, thanks for the issue.

This does look like a bug, and I think I understand what you're trying to do in your pull request. That said, do you happen to have any documentation (etc.) from node-glob/node-ignore which helps explain the changes in your PR (i.e., what requirements on their side we might have missed)? Not strictly required but would help me understand and review the PR better :smile:

**Comment by belochub** [comment]
@platinumazure since https://github.com/eslint/eslint/pull/11546 has landed, I've updated my PR, and node-glob is no longer used. As for node-ignore, one of the conventions specified in the documentation is relevant to this issue: https://github.com/kaelzhang/node-ignore#2-filenames-and-dirnames
